cvmfs/export_plugin/fs_traversal_interface.h

```diff
@@ -0,0 +1,346 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+#ifndef CVMFS_EXPORT_PLUGIN_FS_TRAVERSAL_INTERFACE_H_
+#define CVMFS_EXPORT_PLUGIN_FS_TRAVERSAL_INTERFACE_H_
+
+#include "libcvmfs.h"
+
+#define COPY_BUFFER_SIZE 4096
+
+struct fs_traversal_context {
+  uint64_t version;
+  uint64_t size;
+
+  char *repo;
+  char *base;
+  char *data;
+  char *config;
+  char *lib_version;
+
+  void * ctx;
+};
+
+enum fs_open_type {
+  fs_open_read = 1,
+  fs_open_write = 2,
+  fs_open_append = 4
+};
+
+struct fs_traversal {
+  /**
+   * Local copy of the context that should be set with the
+   * returned value of initialize. Ideally this is created
+   * and then freed all within the traversal.
+   */
+  struct fs_traversal_context *context_;
+
+  /**
+   * Method which initializes a file system traversal context based on
+   * repository and data storage information
+   * 
+   * @param[in] repo Repository name
+   * @param[in] base Base repository location
+   * @param[in] data Data storage information
+   * @param[in] num_threads Number of threads that may be started
+   * @param[in] config A configuration file name
+   * @param[out] A pointer to the context
+   */
+  struct fs_traversal_context *(*initialize)(
+    const char *repo,
+    const char *base,
+    const char *data,
+    int num_threads,
+    const char *config);
+
+  /**
+   * Method which finalizes the file system traversal.
+   * Must always be called after a context had been initialized and before the
+   * procedure ends
+   * 
+   * The context is freed during execution of this method
+   * 
+   * @param[in] ctx The context to finalize and free
+   */
+  void (*finalize)(struct fs_traversal_context *ctx);
+
+  /**
+   * Method that takes the specifications provided and stores
+   * them in a provenance directory. 
+   *
+   * @param[in] src The src context being archived
+   * @param[in] dest The dest context being archived
+   */
+  void (*archive_provenance)(struct fs_traversal_context *src,
```

Comment:
Provenance should be generalized. Possibly specify a directory to write the archive and let the interface decide what to archive. This could make it easier for the interface to read its own provenance and use it. The current implementation leave something to be desired.

Reply:
Also, do we want the contents in the data directory, or written at the file?